lib/puppet/parser/functions/new.rb

```diff
@@ -864,7 +866,7 @@
 | b | The data is in base64 encoding, padding as required by base64 strict is added by default
 | u | The data is in URL safe base64 encoding
 | B | The data is in base64 strict encoding
-| s | The data is a puppet string used verbatim as a binary byte sequence
+| s | The data is a puppet string seen as a binary sequence of bytes with ignored encoding
```

Comment:
The use case is to enable use of Binary instead of just a String (as it is currently used), but where the encoding of the content is not known (it is just binary/raw data - it may be encoded in some text encoding, but that is not known). Ruby does not have a byte buffer, so currently functions return ascii 8 bit strings, or UTF-8 strings. I understand your concern regarding platform independence - would it help if it is specified as "8 bit ascii" (which I believe translates to ISO-8859-1 in Java as it allows any byte values in the string even if it implies that they represent particular glyphs). The alternative is to add the ability to specify the wanted encoding. Those are however platform specific. In any case, a port to some other language would need to revise what functions return - if it should return a Binary or a String and it would need to know when it is doing the reading. If you in Java were to do the stupid thing of creating a string that is really a binary, you could read it as ISO-8859-1.

Once the Binary type is introduced in Puppet we can deprecate / enforce (in some major version) that all puppet strings must be UTF-8, and that Binary is used for everything else.